### PR TITLE
feat: add ranking evaluators

### PR DESCRIPTION
**Pull request by JoanFM**

### CODE REVIEW

File: ranking_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    class DocumentGroundtruthArray(TraversableSequence):
+        def __init__(self, pairs):
+            self._pairs = pairs
+
+        @property
+        def matches(self):
+            pairs = []
+            for doc, groundtruth in zip(self.doc.matches, self.groundtruth.matches):
+                pairs.append(RankingEvaluator.DocGroundtruthPair(doc, groundtruth))
+            return RankingEvaluator.DocumentGroundtruthArray(pairs)
```
Comment (bwanglzu) [inline ranking_evaluator.py]: can we unify `DocGroundtruthPair ` and `DocumentGroundtruthArray `, such as `DocGroundtruthArray`

**Review by bwanglzu — COMMENTED** [review]
i was thinking, maybe we can use the `power_relevance` fomular by default, according to wiki:

>  The latter formula is commonly used in industry including major web search companies[5] and data science competition platforms such as Kaggle

And we do not maintain all the tests/wired argument name for the traditional formulation anymore. How do you think?

**Comment by JoanFM** [comment]
> i was thinking, maybe we can use the `power_relevance` fomular by default, according to wiki:
> 
> > The latter formula is commonly used in industry including major web search companies[5] and data science competition platforms such as Kaggle
> 
> And we do not maintain all the tests/wired argument name for the traditional formulation anymore. How do you think?

I think since it is the default, I would keep it like this. I used to do some experiments about adding this or not. Also maybe some people may have the dataset already precomputed with the `power`

File: ranking_evaluator.py (hunk 1 above)
Comment (JoanFM) [reply ranking_evaluator.py]: Done! Good catch!

File: Dockerfile [hunk 2]
```diff
@@ -1,13 +1,13 @@
 FROM jinaai/jina:master as base
 
-COPY . ./image_normalizer/
-WORKDIR ./image_normalizer
+COPY . ./ranking_evaluator/
```
Comment (nan-wang) [inline Dockerfile:3]: Does this work with `docker build`?
```suggestion
COPY . ./ranking_evaluator/
```

File: Dockerfile [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
+WORKDIR ./ranking_evaluator
 
 RUN pip install .
 
 FROM base
 RUN pip install -r tests/requirements.txt
-RUN pytest tests
+RUN pytest -v -s tests
```
Comment (nan-wang) [inline Dockerfile]: Test is not running in hubble. You can safely remove this part

File: manifest.yml [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
+description: '`RankingEvaluator` exposes different type of evaluation metrics'
 author: Jina AI Dev-Team (dev-team@jina.ai)
 url: https://jina.ai
 vendor: Jina AI Limited
 documentation: https://github.com/jina-ai/jina-hub
 version: 0.0.1
 license: apache-2.0
-keywords: [image, normalization, reader, resize]
\ No newline at end of file
+keywords: [evaluation, ranking, precision, recall, ndcg, reciprocalrank, averageprecision, metric]
```
Comment (nan-wang) [inline manifest.yml]:
```suggestion
keywords: [evaluator, ranker]
```

File: manifest.yml [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
+name: RankingEvaluator
+kind: evaluator
+type: executor
+description: '`RankingEvaluator` exposes different type of evaluation metrics'
 author: Jina AI Dev-Team (dev-team@jina.ai)
 url: https://jina.ai
 vendor: Jina AI Limited
 documentation: https://github.com/jina-ai/jina-hub
 version: 0.0.1
 license: apache-2.0
```
Comment (nan-wang) [inline manifest.yml:11]: check out the keywords at https://github.com/jina-ai/executor-template/blob/main/.github/GUIDELINES.md#how-to-choose-keywords

There are quite a few fields dropped in the Hubble https://github.com/jina-ai/executor-template/blob/main/.github/GUIDELINES.md#how-to-write-manifestyml

File: ranking_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    @property
+    def _evaluation_name(self):
+        return self.evaluation_name or f'{self.metric}@{self.eval_at}' if self.eval_at is not None else self.metric
+
+    @requests
+    def evaluate(self, docs: Optional[DocumentArray], groundtruths: Optional[DocumentArray], parameters: Dict, **kwargs):
+        if docs is None or groundtruths is None:
+            return
+        traversal_pathss = parameters.get('traversal_paths', self.default_traversal_paths)
```
Comment (nan-wang) [inline ranking_evaluator.py]:
```suggestion
        traversal_paths = parameters.get('traversal_paths', self.default_traversal_paths)
```

File: setup.py [hunk 7]
```diff
@@ -2,18 +2,18 @@
 
 
 setuptools.setup(
-    name="executor-image-normalizer",
+    name="executors-evaluator-ranking",
     version="2.0",
```
Comment (nan-wang) [inline setup.py]: This is not JIna version. Instead, it is just the version of the executor.
```suggestion
    version="1",
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (JoanFM) [reply Dockerfile:3]: it works, what is the change requested?

Comment (nan-wang) [reply Dockerfile:3]: my fault. It is ok

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Pull request merged**